### PR TITLE
pcr: add ability to resume replicating after cutting over

### PR DESCRIPTION
**Pull request by dt**
Previously once a virtual cluster completed replication, it could not go back to replicating; if you promoted a replicating virtual cluster to active, you would need to make a _new_ virtual cluster and backfill it entirely if you once again wanted to have a standby virtual cluster replicating from the source.

However a recently promoted former standby still has most of the data that a new cluster would need to backfill, so a much faster option than making a new one, if the activated cluster no longer needs to be activated, is to deactivate it, rewind it to the point at which it diverged, and then resume replcating from there, if still possible.

This change adds the ability to do just that, using the same syntax that can be used in the very similar case of reversing replication where a previously active virtual cluster is also rewound to then be used as replicating standby.

Release note: none.
Epic: none.

### CODE REVIEW

**Comment by dt** [comment]
(Still need to write a couple tests for this, but flushing my local branches to PRs before end of our PCR sprint)

**Comment by dt** [comment]
@stevendanna I think I'm going to remove the last commit from this, that enforces that a PTS cover the revert-to time, until we figure out the PTS windows (or don't). WDYT?

File: pkg/ccl/streamingccl/streamingest/alter_replication_job.go [hunk 1]
```diff
@@ -399,6 +399,48 @@ func alterTenantRestartReplication(
 	), "creating replication job")
 }
 
+func pickReplicationResume(
```
Comment (msbutler) [inline pkg/ccl/streamingccl/streamingest/alter_replication_job.go:402]: i'm finding this function very hard to follow. could you split this into two functions: 1 for resuming a replication stream on the destination tenant after it has cutover; the other for reversing the direction of replication? Once we have the two different functions, I think it will be easier to reason about when we need to throw an error to the user.

Comment (dt) [reply pkg/ccl/streamingccl/streamingest/alter_replication_job.go:402]: These are not actually two distinct cases though. They can both occur at once in which case you want to pick the better one, or neither could be true, in which case you have an error.

Imagine:
T0: A replicating from B.
T1: B runs cuts over.
T2: A starts replicating from B (revert to T1 and resume, aka reversal)
T3: A cuts over.
T4: A resumes replicating from B (from either T1 or T3).

In this case it is both a reversal of a prior replication -- new src's replicatedFrom is the new dest's srcID -- which ended at T1 _and_ a resumption of replication of the same src from which it was previously replicating (until T3), so it can resume from either T3 or T1. T3 is later, so it should use that.

File: pkg/ccl/streamingccl/streamingest/alter_replication_job.go [hunk 2]
```diff
@@ -399,6 +399,48 @@ func alterTenantRestartReplication(
...
+	var resumeTS hlc.Timestamp
+	var wasReplicatingFrom string
+
+	if dstPreviousSrc != nil {
+		// We were previously replicating but have since cutover; if the srcID
+		// we were replicating from matches the the src ID, maybe we can just
+		// restart replicating from when we cutover.
+
+		wasReplicatingFrom = fmt.Sprintf("%s:%s", dstPreviousSrc.ClusterID, dstPreviousSrc.TenantID)
+		if wasReplicatingFrom == srcID {
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/streamingccl/streamingest/alter_replication_job.go:420]: >  shouldn't this error out?
No it just means we cannot use the time at which the destination was previously cutover (so we do _not_ set `resumeTS` to it)

We only have an error if we're both unable to use this timestamp and unable to use the other side's cutover timestamp, meaning we have no timestamp -- which is what's going on if we get to the end of the function and resumeTS is still empty.

File: pkg/ccl/streamingccl/streamingest/alter_replication_job.go (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/streamingccl/streamingest/alter_replication_job.go:402]: I see, and we could have a timeline in which the reversal occurs  after the destination cutover; i.e:

T0: A replicating from B.
T1: B runs cuts over.
T2: A starts replicating from B (revert to T1 and resume, aka reversal)
T3: A cuts over.
T4 : B resumes replicating from T3, instead of T1

 Perhaps for the dull, you could add a docstring to this function describing these two timelines. Further, I think the function could be easier to read if we restructured it to compute `latestCutoverOnCurrentCluster` and `latestCutoverOnSourceCluster`, then return the latest of the two or error if both are empty.

Comment (dt) [reply pkg/ccl/streamingccl/streamingest/alter_replication_job.go:414]: I had some comments inline but I've adjusted them to use the terms reversal and resume explicitly and am keeping the timestamps separate until the end for clarity. WDYT?

File: pkg/ccl/streamingccl/streamclient/client.go [hunk 3]
```diff
@@ -90,7 +90,9 @@ type Client interface {
 	// Complete completes a replication stream consumption.
 	Complete(ctx context.Context, streamID streampb.StreamID, successfulIngestion bool) error
 
-	PriorReplicationDetails(ctx context.Context, tenant roachpb.TenantName) (string, hlc.Timestamp, error)
+	PriorReplicationDetails(
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/streamingccl/streamclient/client.go:102]: I decided "sourceID" and "srcID" are confusing here, since "source" or "src" also is used in contrast to "destination" or "dst" elsewhere in code that handles these IDs. I've revised to to call it a "historyID", regardless of whether it is being used on the src or dst, to reduce the cognitive load a bit. WDYT?

File: pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go [hunk 4]
```diff
@@ -258,26 +259,29 @@ func (p *partitionedStreamClient) Complete(
...
+		d, _, err := apd.NewFromString(activated.String)
+		if err != nil {
+			return "", "", hlc.Timestamp{}, err
+		}
+		ts, err := hlc.DecimalToHLC(d)
+		return curID, srcID.String, ts, err
 	}
-	ts, err := hlc.DecimalToHLC(d)
-	return srcID, ts, err
+	return curID, "", hlc.Timestamp{}, nil
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go]: > return the priorID if the cluster is not activated

Eh, I don't like returning an ID saying it was replicated from an ID if it hasn't finished that replication. For all we know it'll abort and drop or something.

> test both non error return paths of PriorReplicationDetails

Good point, done.

File: pkg/ccl/streamingccl/streamingest/alter_replication_job.go (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/streamingccl/streamingest/alter_replication_job.go:414]: This looks much clearer. thanks for changing!

File: pkg/ccl/streamingccl/streamclient/client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/streamingccl/streamclient/client.go:102]: yeah this explanation works for me.

**Review by msbutler — APPROVED** [review]
Thanks for adding the additional client test!

**Comment by dt** [comment]
TFTR!

bors r+
